src/crimson/os/seastore/scanner.cc

```diff
@@ -0,0 +1,373 @@
+	decode(header, bp);
+      } catch (ceph::buffer::error &e) {
+	return read_validate_record_metadata_ret(
+	  read_validate_record_metadata_ertr::ready_future_marker{},
+	  std::nullopt);
+      }
+      if (header.segment_nonce != nonce) {
+	return read_validate_record_metadata_ret(
+	  read_validate_record_metadata_ertr::ready_future_marker{},
+	  std::nullopt);
+      }
+      if (header.mdlength > (extent_len_t)block_size) {
+	if (start.offset + header.mdlength >
+	    (int64_t)segment_manager.get_segment_size()) {
+	  return crimson::ct_error::input_output_error::make();
+	}
+	return segment_manager.read(
+	  {start.segment, start.offset + (segment_off_t)block_size},
+	  header.mdlength - block_size).safe_then(
+	    [header=std::move(header), bl=std::move(bl)](
+	      auto &&bptail) mutable {
+	      bl.push_back(bptail);
+	      return read_validate_record_metadata_ret(
+		read_validate_record_metadata_ertr::ready_future_marker{},
+		std::make_pair(std::move(header), std::move(bl)));
+	    });
+      } else {
+	return read_validate_record_metadata_ret(
+	  read_validate_record_metadata_ertr::ready_future_marker{},
+	  std::make_pair(std::move(header), std::move(bl))
+	);
+      }
+    }).safe_then([=](auto p) {
+      if (p && validate_metadata(p->second)) {
+	return read_validate_record_metadata_ret(
+	  read_validate_record_metadata_ertr::ready_future_marker{},
+	  std::move(*p)
+	);
+      } else {
+	return read_validate_record_metadata_ret(
+	  read_validate_record_metadata_ertr::ready_future_marker{},
+	  std::nullopt);
+      }
+    });
+}
+
+std::optional<std::vector<extent_info_t>>
+Scanner::try_decode_extent_infos(
+  record_header_t header,
+  const bufferlist &bl)
+{
+  auto bliter = bl.cbegin();
+  bliter += ceph::encoded_sizeof_bounded<record_header_t>();
+  bliter += sizeof(checksum_t) /* crc */;
+  logger().debug("{}: decoding {} extents", __func__, header.extents);
+  std::vector<extent_info_t> extent_infos(header.extents);
+  for (auto &&i : extent_infos) {
+    try {
+      decode(i, bliter);
+    } catch (ceph::buffer::error &e) {
+      return std::nullopt;
+    }
+  }
+  return extent_infos;
+}
+
+Scanner::read_validate_data_ret
+Scanner::read_validate_data(
+  paddr_t record_base,
+  const record_header_t &header)
+{
+  return segment_manager.read(
+    record_base.add_offset(header.mdlength),
+    header.dlength
+  ).safe_then([=, &header](auto bptr) {
+    bufferlist bl;
+    bl.append(bptr);
+    return bl.crc32c(-1) == header.data_crc;
+  });
+}
+
+bool Scanner::validate_metadata(const bufferlist &bl)
+{
+  auto bliter = bl.cbegin();
+  auto test_crc = bliter.crc32c(
+    ceph::encoded_sizeof_bounded<record_header_t>(),
+    -1);
+  ceph_le32 recorded_crc_le;
+  decode(recorded_crc_le, bliter);
+  uint32_t recorded_crc = recorded_crc_le;
+  test_crc = bliter.crc32c(
+    bliter.get_remaining(),
+    test_crc);
+  return test_crc == recorded_crc;
+}
+
+Scanner::init_segments_ret
+Scanner::init_segments() {
```

Comment:
This method should part of SegmentCleaner.  The state changes are entirely within SegmentCleaner.  Scanner doesn't really even need to know how to enumerate segments (entirely under the control of the Segmentmanager for each device).  SegmentCleaner can simply use SegmentManager to enumerate the segments, use this to check the type of each one, and then update its state directly.  That removes the need for Scanner to have a reference to SegmentCleaner.

Reply:
See my other comment, this may actually need to live in EPM/TM.